src/server/engine_factory.rs

```diff
@@ -237,7 +237,7 @@ impl TabletFactory<RocksEngine> for KvEngineFactory {
         RocksEngine::exists(path.to_str().unwrap())
     }
 
-    #[cfg(any(test, feature = "testexport"))]
+    #[cfg(feature = "testexport")]
     fn set_state_storage(&self, state_storage: Arc<dyn StateStorage>) {
         let inner = Arc::as_ptr(&self.inner) as *mut FactoryInner;
         unsafe {
```

Comment:
This can't be more unsafe. The code is broken.

Reply:
Any suggestion?

Reply:
Delete the method and solve it without unsafe. TabletFactory should be recreated for every node when it's started.

Reply:
What potential problems this method could cause? 
Recreating tablet factory when node starts may not be possible according to the current inplementation.

Reply:
Is it possible to make tablet factory mut in the caller side? As it does need the mut-ability.
Or make the state_storage protected by a mutex then you can use non-mut ref in the caller.

Reply:
![image](https://user-images.githubusercontent.com/71589810/222310754-19ef5afd-4d51-45d5-8f13-890a2d0285de.png)
`tablets` is an Arc type and owned by not only one ownner, so to modify it, unsafe method is needed.

And, adding mutex will influence the performance of production code.

Reply:
The whole point of `Arc` is to guarantee the lifetime of data it protects. Using unsafe to overwrite the data breaks the guarantee, it can lead to UAF.